### PR TITLE
Fixed #33047 -- Added constraint support for PostGIS and MySQL GIS backends.

### PR DESCRIPTION
**Pull request by claudep**
Thanks Arsalan Ghassemi for the report and the initial patch.

### CODE REVIEW

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: tests/gis_tests/gis_migrations/test_operations.py [hunk 1]
```diff
@@ -235,6 +235,16 @@ def test_alter_geom_field_dim(self):
         )
         self.assertFalse(Neighborhood.objects.first().geom.hasz)
 
+    def test_add_constraint(self):
+        Neighborhood = self.current_state.apps.get_model('gis', 'Neighborhood')
+        poly = Polygon(((0.0, 0.0), (0.0, 50.0), (50.0, 50.0), (50.0, 0.0), (0.0, 0.0)))
```
Comment (felixxm) [inline tests/gis_tests/gis_migrations/test_operations.py:240]:
```suggestion
        poly = Polygon(((0, 0), (0, 1), (1, 1), (1, 0), (0, 0)))
```

File: tests/gis_tests/gis_migrations/test_operations.py [hunk 2]
```diff
@@ -235,6 +235,16 @@ def test_alter_geom_field_dim(self):
...
+        constraint = models.CheckConstraint(
+            check=models.Q(('geom__within', poly)),
```
Comment (felixxm) [inline tests/gis_tests/gis_migrations/test_operations.py:242]: I would use kwargs
```suggestion
            check=models.Q(geom__within=poly),
```

File: tests/gis_tests/gis_migrations/test_operations.py [hunk 3]
```diff
@@ -235,6 +235,16 @@ def test_alter_geom_field_dim(self):
...
+            name='geom_constraint'
+        )
+        with connection.schema_editor() as editor:
+            editor.add_constraint(Neighborhood, constraint)
+
```
Comment (felixxm) [inline tests/gis_tests/gis_migrations/test_operations.py:247]: Can we check that the constraint actually exists?
```python
        with connection.cursor() as cursor:
            constraints = connection.introspection.get_constraints(
                cursor,
                Neighborhood._meta.db_table,
            )
            self.assertIn(constraint_name, constraints)
```

**Review by felixxm — COMMENTED** [review]
@claudep Thanks for this patch :+1: Unfortunately it doesn't work on Oracle :disappointed:

**Comment by claudep** [comment]
For the Oracle failure, someone else will have to work on on it (or I can skip the test if nobody chimes in soon).

**Comment by claudep** [comment]
I guess I also may need some help for MySQL failures, since tests are passing fine with my local MariaDB instance.

**Comment by felixxm** [comment]
> I guess I also may need some help for MySQL failures, since tests are passing fine with my local MariaDB instance.

I fixed tests failures on MySQL < 8.0.16 and will try to figure out how to fix this on Oracle.

**Comment by claudep** [comment]
Thanks a lot!

File: django/contrib/gis/db/backends/oracle/features.py [hunk 4]
```diff
@@ -12,3 +13,13 @@ class DatabaseFeatures(BaseSpatialFeatures, OracleDatabaseFeatures):
...
     supports_tolerance_parameter = True
     unsupported_geojson_options = {'bbox', 'crs', 'precision'}
+
+    @cached_property
+    def django_test_skips(self):
+        skips = super().django_test_skips
+        skips.update({
+            "Oracle doesn't support spatial operators in constraints.": {
+                'gis_tests.gis_migrations.test_operations.OperationTests.test_add_check_constraint',
+            },
```
Comment (felixxm) [inline django/contrib/gis/db/backends/oracle/features.py:23]: It seems that spatial operators cannot be used in constraints, see [_"Usage notes"_](https://docs.oracle.com/en/database/oracle/oracle-database/19/spatl/spatial-operators-reference.html#GUID-97C17C18-F05E-49B4-BE11-E89B972E2A02):

> _"The operator must always be used in a **WHERE** clause, ..."_

@jtiai Jani, Can you take a look?

File: django/contrib/gis/db/backends/oracle/schema.py [hunk 5]
```diff
@@ -31,6 +31,11 @@ def __init__(self, *args, **kwargs):
     def geo_quote_name(self, name):
         return self.connection.ops.geo_quote_name(name)
 
+    def quote_value(self, value):
+        if isinstance(value, self.connection.ops.Adapter):
+            return super().quote_value(str(value))
+        return super().quote_value(value)
```
Comment (felixxm) [inline django/contrib/gis/db/backends/oracle/schema.py:37]: I decided to fix `quote_value()` anyway, in order not to mask the original exception:
```
ORA-00904: "SDO_EQUAL": invalid identifier
```
with
```
ORA-00907: missing right parenthesis
```

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: django/contrib/gis/db/backends/oracle/features.py (hunk 4 above)
Comment (jtiai) [reply django/contrib/gis/db/backends/oracle/features.py:23]: I think that only applies to `SDO_FILTER` but I never recall seeing operators being used in constraints.

Comment (felixxm) [reply django/contrib/gis/db/backends/oracle/features.py:23]: I tried to use a generated condition the `WHERE` clause and it works fine, so IMO we can safely assume that it's an Oracle caveat :smile:

Comment (felixxm) [reply django/contrib/gis/db/backends/oracle/features.py:23]: Thanks!

**Pull request merged**